File: tools/extract_strings.py
#!/usr/bin/env python3
"""
Extractor voor zichtbare NL-strings uit de NL-app.
Output: JSON met {sleutel: {nl, bestand, regel, context}}.

Sleutel-conventie: bestandsnaam_zonder_ext.slug_van_eerste_woorden
Voorbeeld: chrome.stemgedrag_app_openvizier

Alleen echt zichtbare UI-tekst wordt geëxtraheerd:
- JSX-tekst tussen > en <
- Quoted strings in props: title=, placeholder=, label=, ariaLabel=, alt=
- Return statements van hulpfuncties die tekst leveren

We negeren:
- import/export paden
- className/id/data-testid/href/src waarden
- console.log, throw new Error (technische strings)
- comments
- shadcn/ui/ map (taal-neutrale gegenereerde code)
"""
from __future__ import annotations
import json
import re
import sys
from pathlib import Path
from unicodedata import normalize

SRC = Path("/home/user/workspace/nl_app/client/src")

# Alleen app-eigen bestanden, niet shadcn/ui
KERN_BESTANDEN = [
    "components/Chrome.tsx",
    "components/MatrixSectie.tsx",
    "components/HistorischeTrend.tsx",
    "components/LevensloopGrafiek.tsx",
    "components/GezondheidsgrafiekNL.tsx",
    "components/PartijDetail.tsx",
    "components/MiniHero.tsx",
    "components/PersoonlijkeWeging.tsx",
    "components/Logo.tsx",
    "pages/Hoofdpagina.tsx",
    "pages/Methodologie.tsx",
    "pages/PersonaFlow.tsx",
    "pages/Sector.tsx",
    "pages/Kaart.tsx",
    "pages/not-found.tsx",
    "lib/personaEngine.ts",
    "lib/pdf.ts",
    "lib/levensloopEngine.ts",
]

# Regex-patronen voor zichtbare tekst
PATRONEN = [
    # JSX-tekst: >Nederlands hier<
    (re.compile(r">([^<>{}\n]+?)<"), "jsx"),
    # Prop-strings: title="...", placeholder="...", label="..."
    (re.compile(r'\b(?:title|placeholder|label|ariaLabel|aria-label|alt|description|subtitle|caption)=["\']([^"\'<>]+?)["\']'), "prop"),
    # Toast / alert titles in objects: title: "...", description: "..."
    (re.compile(r'\b(?:title|description|message|label|naam|beschrijving)\s*:\s*["\']([^"\'<>]+?)["\']'), "obj"),
]

# NL-markers om te detecteren of een string echt NL is
NL_MARKERS = re.compile(
    r'\b('
    r'de|het|een|is|zijn|wordt|worden|niet|geen|voor|met|van|naar|door|op|in|uit|aan|bij|om|te|dat|die|dit|deze|wat|hoe|waarom|waar|wanneer|wie|welk|welke|jouw|uw|jullie|onze|hun|ook|maar|of|als|dan|zo|nog|al|zeer|erg|heel|meer|minder|veel|weinig|elk|elke|geen|alleen|samen|zelf|eigen|nieuw|oud|goed|slecht|groot|klein'
    r')\b',
    re.IGNORECASE,
)

# Skip als een van deze patronen voorkomt
SKIP_INHOUD = re.compile(r'^(https?://|/|#|\{|\.|,|\s|-|\+|=|\d|[a-z_]+\(|null|true|false|undefined)$|^[A-Z_]+$', re.IGNORECASE)


def slugify(tekst: str, max_len: int = 30) -> str:
    """Maak stabiele sleutel-slug van tekst."""
    tekst = normalize("NFKD", tekst).encode("ascii", "ignore").decode("ascii")
    tekst = re.sub(r"[^a-zA-Z0-9\s]", "", tekst).lower()
    woorden = tekst.split()[:5]
    slug = "_".join(woorden)[:max_len].strip("_")
    return slug or "leeg"


def is_zichtbare_nl(tekst: str) -> bool:
    """Beslissing: is dit echt zichtbare NL-tekst?"""
    tekst = tekst.strip()
    if not tekst or len(tekst) < 3:
        return False
    if len(tekst) > 500:
        return False
    if SKIP_INHOUD.match(tekst):
        return False
    # Skip HTML entities alleen
    if re.fullmatch(r"(&[a-z]+;|\s)+", tekst):
        return False
    # Skip als het pure code lijkt (veel {}, JS-operatoren)
    if tekst.count("{") + tekst.count("}") > 2:
        return False
    if re.search(r'=>|===|!==|&&|\|\|', tekst):
        return False
    # Minstens 1 NL-marker of duidelijk NL-teken (accenten, hoofdletter-gevolgd-door-tekst)
    if NL_MARKERS.search(tekst):
        return True
    # Losse hoofdletter-woorden ("Volgende", "Terug") — accepteren als >4 letters
    if re.fullmatch(r"[A-ZÀ-ÿ][a-zà-ÿ]{3,}(\s[A-Za-zÀ-ÿ]+)*[.!?]?", tekst):
        return True
    return False


def extract_uit_bestand(pad: Path) -> list[dict]:
    """Retourneer lijst van {sleutel, nl, regel, context, patroon}."""
    if not pad.exists():
        return []
    inhoud = pad.read_text(encoding="utf-8")
    regels = inhoud.splitlines()
    modulenaam = pad.stem.lower().replace("-", "_")

    hits: list[dict] = []
    gezien: set[str] = set()

    for regelnr, regel in enumerate(regels, 1):
        # Skip pure comment-regels
        gestript = regel.strip()
        if gestript.startswith("//") or gestript.startswith("*") or gestript.startswith("/*"):
            continue
        # Skip import/export regels
        if re.match(r"^\s*(import|export\s+(default\s+)?(from|type|\*|\{))", gestript):
            continue

        for patroon, ptype in PATRONEN:
            for match in patroon.finditer(regel):
                tekst = match.group(1).strip()
                # Ontsnappen van HTML entities lezen we niet — laat ze staan zoals ze in de tsx-bron staan
                if not is_zichtbare_nl(tekst):
                    continue
                if tekst in gezien:
                    continue
                gezien.add(tekst)
                sleutel = f"{modulenaam}.{slugify(tekst)}"
                # Uniek maken bij collision
                basis_sleutel = sleutel
                teller = 2
                while any(h["sleutel"] == sleutel for h in hits):
                    sleutel = f"{basis_sleutel}_{teller}"
                    teller += 1
                hits.append({
                    "sleutel": sleutel,
                    "nl": tekst,
                    "bestand": str(pad.relative_to(SRC)),
                    "regel": regelnr,
                    "patroon": ptype,
                })
    return hits


def main() -> None:
    alle: list[dict] = []
    for rel in KERN_BESTANDEN:
        pad = SRC / rel
        hits = extract_uit_bestand(pad)
        alle.extend(hits)
        print(f"  {rel}: {len(hits)} strings", file=sys.stderr)
    print(f"\nTotaal: {len(alle)} unieke NL-strings", file=sys.stderr)
    print(json.dumps(alle, ensure_ascii=False, indent=2))


if __name__ == "__main__":
    main()

File: tools/extract_strings.mjs
#!/usr/bin/env node
/**
 * AST-based extractor voor zichtbare NL-strings uit de NL-app.
 * Gebruikt @babel/parser + @babel/traverse.
 *
 * Uitvoer: JSON-lijst {sleutel, nl, bestand, regel, kolom, type}
 *
 * Extractie-regels:
 *   - JSXText knopen (tekst tussen JSX-tags) → altijd meenemen als niet-leeg
 *   - JSXAttribute met StringLiteral value, alleen voor tekst-attributen:
 *       title, placeholder, label, aria-label, ariaLabel, alt, description
 *   - ObjectProperty met StringLiteral value, alleen voor tekst-keys:
 *       title, description, label, message, naam, beschrijving, subtitle,
 *       caption, text, summary, hint, error, waarschuwing
 *   - TemplateLiteral quasi's (statische delen) alleen als ze in tekst-context staan
 *
 * Filter: alleen "NL-achtige" strings (min. 3 tekens, geen pure code, bevat NL-marker of hoofdletter-gestart tekstpatroon).
 */

import fs from 'node:fs';
import path from 'node:path';
import { parse } from '@babel/parser';
import traverseModule from '@babel/traverse';

const traverse = traverseModule.default;

const SRC = '/home/user/workspace/nl_app/client/src';
const BESTANDEN = [
  'components/Chrome.tsx',
  'components/MatrixSectie.tsx',
  'components/HistorischeTrend.tsx',
  'components/LevensloopGrafiek.tsx',
  'components/GezondheidsgrafiekNL.tsx',
  'components/PartijDetail.tsx',
  'components/MiniHero.tsx',
  'components/PersoonlijkeWeging.tsx',
  'components/Logo.tsx',
  'pages/Hoofdpagina.tsx',
  'pages/Methodologie.tsx',
  'pages/PersonaFlow.tsx',
  'pages/Sector.tsx',
  'pages/Kaart.tsx',
  'pages/not-found.tsx',
  'lib/personaEngine.ts',
  'lib/pdf.ts',
  'lib/levensloopEngine.ts',
];

const TEKST_ATTRIBUTEN = new Set([
  'title', 'placeholder', 'label', 'aria-label', 'ariaLabel', 'alt',
  'description', 'subtitle', 'caption', 'text', 'hint', 'summary',
]);

const TEKST_OBJECT_KEYS = new Set([
  'title', 'description', 'label', 'message', 'naam', 'beschrijving',
  'subtitle', 'caption', 'text', 'summary', 'hint', 'error', 'waarschuwing',
  'kop', 'toelichting', 'uitleg', 'vraag', 'antwoord', 'optie',
  'kort', 'lang', 'context', 'tooltip', 'note', 'notitie', 'melding',
]);

const NL_MARKERS = /\b(de|het|een|is|zijn|wordt|worden|niet|geen|voor|met|van|naar|door|op|in|uit|aan|bij|om|te|dat|die|dit|deze|wat|hoe|waarom|waar|wanneer|wie|welk|welke|jouw|uw|jullie|onze|hun|ook|maar|of|als|dan|zo|nog|al|zeer|erg|heel|meer|minder|veel|weinig|elk|elke|alleen|samen|zelf|eigen|nieuw|oud|goed|slecht|groot|klein|per|tot|over|onder|tegen|zonder|volgens|tijdens)\b/i;

function isZichtbareNL(tekst) {
  tekst = tekst.trim();
  if (!tekst || tekst.length < 3 || tekst.length > 800) return false;
  // Skip URLs, paden, technische strings
  if (/^https?:\/\//.test(tekst)) return false;
  if (/^\/[a-z\-\/]+$/i.test(tekst)) return false;
  if (/^#[a-z\-]+$/i.test(tekst)) return false;
  if (/^[a-z_][a-zA-Z0-9_]*$/.test(tekst) && !/\s/.test(tekst)) return false; // identifiers
  if (/^[A-Z_]+$/.test(tekst)) return false; // CONSTANTEN
  if (/^\d+(\.\d+)?%?$/.test(tekst)) return false; // pure getallen
  // Skip als het overduidelijk code is
  if (/=>|===|!==|&&|\|\|/.test(tekst)) return false;
  // Skip CSS class-lijstjes: veel dashes, alle kleine letters, geen echte spaties tussen woorden
  if (/^[a-z0-9\-\s:]+$/.test(tekst) && /-/.test(tekst) && !NL_MARKERS.test(tekst)) return false;
  // Skip HTML entity-only strings
  if (/^(&[a-z]+;|\s)+$/.test(tekst)) return false;

  // Positieve tests
  if (NL_MARKERS.test(tekst)) return true;
  // Hoofdlettergestart tekst met minstens 4 tekens
  if (/^[A-ZÀ-ÿ][a-zA-ZÀ-ÿ\s.,!?'\-]{3,}$/.test(tekst) && /[a-zà-ÿ]/.test(tekst)) return true;
  // Bevat expliciet Nederlandse accenten of ij
  if (/[ë][a-z]|ij[a-z]|[oe][iu]/.test(tekst) && /[a-zà-ÿ]{4,}/.test(tekst)) return true;
  return false;
}

function slugify(tekst, maxLen = 35) {
  const zonderAccenten = tekst.normalize('NFKD').replace(/[\u0300-\u036f]/g, '');
  const genormaliseerd = zonderAccenten
    .replace(/&[a-z]+;/g, ' ')
    .replace(/[^a-zA-Z0-9\s]/g, ' ')
    .toLowerCase()
    .trim();
  const woorden = genormaliseerd.split(/\s+/).slice(0, 6);
  const slug = woorden.join('_').slice(0, maxLen).replace(/_+$/, '');
  return slug || 'leeg';
}

function attrName(node) {
  if (!node) return null;
  if (node.type === 'JSXIdentifier') return node.name;
  if (node.type === 'JSXNamespacedName') return `${node.namespace.name}:${node.name.name}`;
  return null;
}

function extractUitBestand(bestandsPad) {
  const volledig = path.join(SRC, bestandsPad);
  if (!fs.existsSync(volledig)) return [];
  const bron = fs.readFileSync(volledig, 'utf8');
  const modulenaam = path.basename(bestandsPad, path.extname(bestandsPad))
    .toLowerCase().replace(/[^a-z0-9]/g, '_');

  let ast;
  try {
    ast = parse(bron, {
      sourceType: 'module',
      plugins: ['typescript', 'jsx'],
      errorRecovery: true,
    });
  } catch (err) {
    console.error(`[FOUT] Parse ${bestandsPad}: ${err.message}`);
    return [];
  }

  const hits = [];
  const gezien = new Set();

  function push(tekst, loc, type) {
    tekst = tekst.replace(/\s+/g, ' ').trim();
    if (!isZichtbareNL(tekst)) return;
    if (gezien.has(tekst)) return;
    gezien.add(tekst);
    const basis = `${modulenaam}.${slugify(tekst)}`;
    let sleutel = basis;
    let teller = 2;
    while (hits.some(h => h.sleutel === sleutel)) {
      sleutel = `${basis}_${teller++}`;
    }
    hits.push({
      sleutel,
      nl: tekst,
      bestand: bestandsPad,
      regel: loc?.start?.line ?? 0,
      kolom: loc?.start?.column ?? 0,
      type,
    });
  }

  traverse(ast, {
    JSXText(pad) {
      push(pad.node.value, pad.node.loc, 'jsx_text');
    },
    JSXAttribute(pad) {
      const naam = attrName(pad.node.name);
      if (!naam || !TEKST_ATTRIBUTEN.has(naam)) return;
      const val = pad.node.value;
      if (val && val.type === 'StringLiteral') {
        push(val.value, val.loc, `attr_${naam}`);
      }
      if (val && val.type === 'JSXExpressionContainer' && val.expression.type === 'StringLiteral') {
        push(val.expression.value, val.expression.loc, `attr_${naam}`);
      }
    },
    ObjectProperty(pad) {
      const key = pad.node.key;
      const keyNaam = key.type === 'Identifier' ? key.name
        : key.type === 'StringLiteral' ? key.value
        : null;
      if (!keyNaam || !TEKST_OBJECT_KEYS.has(keyNaam)) return;
      const val = pad.node.value;
      if (val.type === 'StringLiteral') {
        push(val.value, val.loc, `obj_${keyNaam}`);
      }
      if (val.type === 'TemplateLiteral' && val.expressions.length === 0 && val.quasis.length === 1) {
        push(val.quasis[0].value.cooked, val.loc, `obj_${keyNaam}`);
      }
    },
    // Return statements van pure tekst-strings in functies (bijv. hulpfuncties)
    ReturnStatement(pad) {
      const arg = pad.node.argument;
      if (arg?.type === 'StringLiteral') {
        push(arg.value, arg.loc, 'return_string');
      }
      if (arg?.type === 'TemplateLiteral' && arg.expressions.length === 0 && arg.quasis.length === 1) {
        push(arg.quasis[0].value.cooked, arg.loc, 'return_template');
      }
    },
  });

  return hits;
}

const alle = [];
for (const rel of BESTANDEN) {
  const hits = extractUitBestand(rel);
  console.error(`  ${rel}: ${hits.length} strings`);
  alle.push(...hits);
}
console.error(`\nTotaal: ${alle.length} unieke NL-strings`);
process.stdout.write(JSON.stringify(alle, null, 2));
